runtimevar/runtimevar.go

```diff
@@ -42,32 +43,76 @@ type Snapshot struct {
 
 // Variable provides the ability to read runtime variables with its blocking Watch method.
 type Variable struct {
-	watcher  driver.Watcher
-	nextCall time.Time
-	prev     driver.State
+	watcher    driver.Watcher
+	nextCall   time.Time
+	prev       driver.State
+	ctx        context.Context
+	cancel     func()
+	initLatest sync.Once
+
+	mu       sync.Mutex
+	lastGood interface{}
 }
 
 // New constructs a Variable object given a driver.Watcher implementation.
 func New(w driver.Watcher) *Variable {
-	return &Variable{watcher: w}
+	// ctx/cancel are used to manage the background goroutine in Latest.
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Variable{
+		watcher: w,
+		ctx:     ctx,
+		cancel:  cancel,
+	}
 }
 
-// Watch blocks until there are variable changes, the Context's Done channel
-// closes or a new error occurs.
+// Latest returns the latest successfully retrieved value of the variable.
+//
+// The first call to Latest will block until the variable is successfully
+// retrieved. Call Close to abort the call; if aborted, Latest will return nil.
 //
-// If the variable changes, the method returns a Snapshot object containing the
-// updated value.
+// Subsequently, Latest will return the latest successfully retrieved value of
+// the variable without blocking. Watch will be called in a background
+// goroutine to receive variable updates. Errors returned by Watch are ignored
+// in favor of returning the latest "good" value.
 //
-// If method returns an error, the returned Snapshot object is a zero value and cannot be used.
+// Applications should call Latest or Watch, not both.
 //
-// The first call to this method should return the current variable unless there are errors in
-// retrieving the value.
+// Latest is safe to be called from multiple goroutines.
+func (c *Variable) Latest() interface{} {
+	c.mu.Lock()
+	defer c.mu.Unlock()
+
+	c.initLatest.Do(func() {
+		// The first time, block until we get a good value.
+		for c.ctx.Err() == nil && c.lastGood == nil {
+			s, _ := c.Watch(c.ctx)
+			if s.Value != nil {
+				c.lastGood = s.Value
+			}
+		}
+		// ... and spin up a goroutine to Watch for updates.
+		go func() {
+			for c.ctx.Err() == nil {
+				s, _ := c.Watch(c.ctx)
+				if s.Value != nil {
+					c.mu.Lock()
+					c.lastGood = s.Value
+					c.mu.Unlock()
+				}
+			}
+		}()
+	})
+	return c.lastGood
+}
+
+// Watch blocks until there are variable changes, the Context's Done channel
+// closes or a new error occurs. It returns a Snapshot containing the variable's
+// Value, or an error.
 //
-// Users should not call this method from multiple goroutines as implementations may not guarantee
-// safety in data access.  It is typical to use only one goroutine calling this method in a loop.
+// Applications should call Latest or Watch, not both.
```

Comment:
If this is the case, then there should be a runtime signifier of this being an invalid state, or even better, an anti-affordance to prevent this from happening.

Reply:
It now panics.